kafka/consumer.go

```diff
@@ -149,31 +151,44 @@ func NewConsumer(cfg ConsumerConfig) (*Consumer, error) {
 		cfg:      cfg,
 		client:   client,
 		consumer: consumer,
+		closed:   make(chan struct{}),
+		running:  make(chan struct{}),
 		tracer:   cfg.tracerProvider().Tracer("kafka"),
 	}, nil
 }
 
 // Close the consumer, blocking until all partition consumers are stopped.
 func (c *Consumer) Close() error {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	// Since we take the full lock when closing, there's no need to explicitly
-	// allow rebalances since polls aren't concurrent with Close().
-	c.client.Close()
-	c.consumer.wg.Wait() // Wait for all the goroutines to exit.
+	select {
+	case <-c.closed:
+	default:
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.client.Close()
+		c.consumer.wg.Wait() // Wait for all the goroutines to exit.
+		close(c.closed)
+	}
 	return nil
 }
 
 // Run the consumer until a non recoverable error is found:
-//   - context.Cancelled.
-//   - context.DeadlineExceeded.
-//   - kgo.ErrClientClosed.
 //   - ErrCommitFailed.
 //
 // To shut down the consumer, cancel the context, or call consumer.Close().
+// If called more than once, returns `apmqueue.ErrConsumerAlreadyRunning`.
 func (c *Consumer) Run(ctx context.Context) error {
+	select {
+	case <-c.running:
+		return apmqueue.ErrConsumerAlreadyRunning
+	default:
+		close(c.running)
+	}
```

Comment:
This can lead to panic due to multiple close on a channel if `Run` is called multiple times in quick succession. As an alternative what do you think about using `atomic.Bool` with its `CompareAndSwap` operation:

```suggestion
        // c.running is defined as an atomic.Bool
	if !c.running.CompareAndSwap(false, true) {
	        return apmqueue.ErrConsumerAlreadyRunning
	}
```